File: tests/fixtures/ecosystems/perl_comprehensive/classes.pl
#!/usr/bin/perl
use strict;
use warnings;

package Animal;

sub new {
    my ($class, %args) = @_;
    return bless {
        name    => $args{name} || 'Unknown',
        species => $args{species} || 'Unknown',
    }, $class;
}

sub name { return $_[0]->{name} }
sub species { return $_[0]->{species} }

sub speak {
    my ($self) = @_;
    return $self->{name} . " makes a sound";
}

package Dog;
use parent -norequire, 'Animal';

sub new {
    my ($class, %args) = @_;
    $args{species} = 'Canine';
    return $class->SUPER::new(%args);
}

sub speak {
    my ($self) = @_;
    return $self->{name} . " barks";
}

sub fetch {
    my ($self, $item) = @_;
    return $self->{name} . " fetches $item";
}

package Cat;
use parent -norequire, 'Animal';

sub new {
    my ($class, %args) = @_;
    $args{species} = 'Feline';
    return $class->SUPER::new(%args);
}

sub speak {
    my ($self) = @_;
    return $self->{name} . " meows";
}

1;

File: tests/fixtures/ecosystems/javascript_comprehensive/classes.js
/**
 * Class patterns for parser testing.
 */

class Animal {
    #name;

    constructor(name, species) {
        this.#name = name;
        this.species = species;
    }

    get name() {
        return this.#name;
    }

    set name(value) {
        this.#name = value;
    }

    speak() {
        return `${this.#name} makes a sound`;
    }

    static create(name, species) {
        return new Animal(name, species);
    }
}

class Dog extends Animal {
    constructor(name) {
        super(name, 'Canine');
    }

    speak() {
        return `${this.name} barks`;
    }

    fetch(item) {
        return `${this.name} fetches ${item}`;
    }
}

class Cat extends Animal {
    constructor(name) {
        super(name, 'Feline');
    }

    speak() {
        return `${this.name} meows`;
    }
}

module.exports = { Animal, Dog, Cat };
